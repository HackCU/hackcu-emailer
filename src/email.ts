import { Ora } from 'ora';
import sendGrid from '@sendgrid/mail';
import { generateTemplate } from './template';

interface Company {
  name: string;
  email: string;
}

export async function sendMassEmail(
  info: { name: string; role: string; emailTo: string },
  companies: Company[],
  {
    apiKey,
    verbose,
    send,
  }: {
    apiKey: string;
    verbose: boolean;
    send: boolean;
  },
  spinner?: Ora,
): Promise<{ worked: Company[]; failed: Company[] }> {
  const worked: Company[] = [];
  const failed: Company[] = [];

  if (send) {
    sendGrid.setApiKey(apiKey);
  }

  for (const company of companies) {
    spinner?.start(`Sending email to "${company.name}" with email "${company.email}".`);

    if (send) {
      try {
        await sendGrid.send({
          to: company.email,
          from: {
            name: 'HackCU',
            email: 'contact@hackcu.org',
          },
          subject: 'Sponsoring HackCU',
          html: generateTemplate({
            company: company.name,
            name: info.name,
            role: info.role,
            emailTo: info.emailTo,
          }),
        });
        spinner?.succeed(`Sent email to "${company.name}" with email "${company.email}"`);
        worked.push(company);
      } catch (error) {
        console.error(error);
        spinner?.fail(`Failed to send email to "${company.name}" with email "${company.email}"`);
        failed.push(company);
      }
    } else {
      spinner?.succeed(`Sent email to "${company.name}" with email "${company.email}"`);
    }
  }

  return { worked, failed };
}
