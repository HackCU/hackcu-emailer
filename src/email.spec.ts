jest.mock('@sendgrid/mail', () => ({
  setApiKey: jest.fn(),
  send: jest.fn(),
}));
import sendGrid from '@sendgrid/mail';
import { sendMassEmail } from './email';

const info = {
  name: 'Kyle',
  role: 'Tech Director',
  emailTo: 'contact@hackcu.org',
};
const apiKey = 'api-key';
const companies = [
  {
    name: 'Company A',
    email: 'companyA@gmail.com',
  },
  {
    name: 'Company B',
    email: 'companyB@gmail.com',
  },
  {
    name: 'Company C',
    email: 'companyC@gmail.com',
  },
];

describe('sendMassEmail', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });
  it('sends emails to all companies', async () => {
    expect.assertions(3);

    (sendGrid.send as jest.Mock).mockResolvedValue(undefined);

    await sendMassEmail(info, companies, {
      apiKey,
      verbose: false,
      send: true,
    });

    expect(sendGrid.setApiKey).toHaveBeenCalledWith(apiKey);
    expect(sendGrid.send).toHaveBeenCalledTimes(companies.length);
    expect((sendGrid.send as jest.Mock).mock.calls).toMatchSnapshot();
  });

  it('returns successful emails and failed emails for companies', async () => {
    expect.assertions(2);

    // Ignore console.error values
    jest.spyOn(console, 'error').mockReturnValue(undefined);

    // Works for first
    (sendGrid.send as jest.Mock).mockResolvedValueOnce(undefined);
    // Fails for second
    (sendGrid.send as jest.Mock).mockRejectedValueOnce(new Error('Network Error'));
    // Works for third
    (sendGrid.send as jest.Mock).mockResolvedValueOnce(undefined);

    const { worked, failed } = await sendMassEmail(info, companies, {
      apiKey,
      verbose: false,
      send: true,
    });

    expect(worked).toEqual([companies[0], companies[2]]);
    expect(failed).toEqual([companies[1]]);
  });
});
