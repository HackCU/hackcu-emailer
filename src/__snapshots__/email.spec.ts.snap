// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendMassEmail sends emails to all companies 1`] = `
Array [
  Array [
    Object {
      "from": Object {
        "email": "contact@hackcu.org",
        "name": "HackCU",
      },
      "html": "<div><p>Hello Company A,</p><p>My name is Kyle and I&#x27;m currently a student studying at the University of Colorado, Boulder. I&#x27;m also an organizer of HackCU, a student run organization that is committed to bringing high quality, inclusive technology and design events including some of Colorado&#x27;s largest student hackathons. We would love to have Company A be a sponsor this year!</p><p>In March, we will be hosting HackCU 007, our main hackathon! HackCU 007 is our first virtual hackathon for 2021 where most passionate student developers will come together to build things such as mobile apps, websites, hardware devices, and other cool stuff to show off their talent and skills!</p><p>You can find more information about HackCU here: <a href=\\"https://hackcu.org/\\">https://hackcu.org/</a></p><p>Over the past few years, we have brought together some of the most gifted and innovative minds in an inspiring and fun setting and are excited to continue to bring sponsorship packages for our event. Feel free to reach out to us if you have any questions. We look forward to hearing back from you!</p><p>Best, <br/>Kyle<br/>Tech Director<br/>contact@hackcu.org</p></div>",
      "subject": "Sponsoring HackCU",
      "to": "companyA@gmail.com",
    },
  ],
  Array [
    Object {
      "from": Object {
        "email": "contact@hackcu.org",
        "name": "HackCU",
      },
      "html": "<div><p>Hello Company B,</p><p>My name is Kyle and I&#x27;m currently a student studying at the University of Colorado, Boulder. I&#x27;m also an organizer of HackCU, a student run organization that is committed to bringing high quality, inclusive technology and design events including some of Colorado&#x27;s largest student hackathons. We would love to have Company B be a sponsor this year!</p><p>In March, we will be hosting HackCU 007, our main hackathon! HackCU 007 is our first virtual hackathon for 2021 where most passionate student developers will come together to build things such as mobile apps, websites, hardware devices, and other cool stuff to show off their talent and skills!</p><p>You can find more information about HackCU here: <a href=\\"https://hackcu.org/\\">https://hackcu.org/</a></p><p>Over the past few years, we have brought together some of the most gifted and innovative minds in an inspiring and fun setting and are excited to continue to bring sponsorship packages for our event. Feel free to reach out to us if you have any questions. We look forward to hearing back from you!</p><p>Best, <br/>Kyle<br/>Tech Director<br/>contact@hackcu.org</p></div>",
      "subject": "Sponsoring HackCU",
      "to": "companyB@gmail.com",
    },
  ],
  Array [
    Object {
      "from": Object {
        "email": "contact@hackcu.org",
        "name": "HackCU",
      },
      "html": "<div><p>Hello Company C,</p><p>My name is Kyle and I&#x27;m currently a student studying at the University of Colorado, Boulder. I&#x27;m also an organizer of HackCU, a student run organization that is committed to bringing high quality, inclusive technology and design events including some of Colorado&#x27;s largest student hackathons. We would love to have Company C be a sponsor this year!</p><p>In March, we will be hosting HackCU 007, our main hackathon! HackCU 007 is our first virtual hackathon for 2021 where most passionate student developers will come together to build things such as mobile apps, websites, hardware devices, and other cool stuff to show off their talent and skills!</p><p>You can find more information about HackCU here: <a href=\\"https://hackcu.org/\\">https://hackcu.org/</a></p><p>Over the past few years, we have brought together some of the most gifted and innovative minds in an inspiring and fun setting and are excited to continue to bring sponsorship packages for our event. Feel free to reach out to us if you have any questions. We look forward to hearing back from you!</p><p>Best, <br/>Kyle<br/>Tech Director<br/>contact@hackcu.org</p></div>",
      "subject": "Sponsoring HackCU",
      "to": "companyC@gmail.com",
    },
  ],
]
`;
